enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,476 @@
+Only *dot* and *bracket* path notation is supported, no wildcards.
+
+For example, given this input record
+```json
+{
+  "kubernetes": {
+    "namespace_name":"foo",
+    "labels": { "kubernetes.io/app": "bar", "other": "thing"},
+  },
+  "hostname": "bart",
+  "level": "critical",
+  ... other meta-data
+  "structured": { "hello": "world" }
+}
+```
+
+this pipeline
+``` yaml
+pipeline:
+- modify:
+    keep: [hostname, level]
+    copy:
+    - kubernetes.namespace_name: my_namespace
+    - "kubernetes.labels['kubernetes.io/app']": my_app # Bracket notation for name with '.'.
+    merge: true
+```
+
+would produce this output record
+```json
+{
+  "hello": "world"
+  "my_namespace": "foo",
+  "my_app": "bar,"
+  "hostname": "bart",
+  "level": "critical",
+}
+```
+
+No change to `output`. Some outputs (e.g. syslog) have additional options for constructing the output record.
+Those rules still apply to the output records modified by the `pipeline.filters`.
+
+### User Stories
+
+Imagine there are two web server implementations named "patchy" and "bulldog".
+Each type of web server produces structured logs in a consistent format, but each uses a different format.
+A hypothetical admin deploys both patchy and bulldog servers.
+
+#### I want to index logs from applications with different schema in Elasticsearch
+
+In both examples below, Elasticsearch will create separate indices for each schema "patchy and "bulldog".
+See the [Elasticsearch notes](#es-notes) for more details.
+
+**Note**: If there is only one JSON schema, the `schema` field is not necessary.
+
+##### Example 1: Identify applications by namespace.
+
+Pods in namespaces "fred" and "bob" use the "patchy" schema, pods in namespaces "jill" and "jane" use the "bulldog" schema.
+
+**Note**: the `namespaces` selector below is already implemented by the CLF.
+
+```yaml
+inputs:
+- name: InputPatchy
+  application:
+    namespaces: [ fred, bob ]
+  structured: json
+  schema: patchy
+- name: InputBulldog
+  application:
+    namespaces: [ jill, jane ]
+  structured: json
+  schema: bulldog
+pipeline:
+- inputRefs: [ InputPatchy, InputBulldog ]
+  outputRefs: default
+```
+
+##### Example 2: Identify applications by labels.
+
+The applications using each schema must be identified by k8s labels.
+Note this is often _already done_ as part of a Deployment.
+Deployments usually label pods by the application they run, which determines their log format.
+
+**Note**: the label `selector` below is defined by the [forwarder label selectors proposal][clf-labels]
+
+```yaml
+inputs:
+- name: InputPatchy
+  applications:
+    selector:
+      matchLabels: { app: patchy }
+  structured: json
+  schema: patchy
+- name: InputBulldog
+  applications:
+    selector:
+      matchLabels: { app: bulldog }
+  structured: json
+  schema: bulldog
+pipelines:
+- inputRefs: [ InputPatchy, InputBulldog ]
+  outputRefs: default
+```
+
+#### I want to forward structured and unstructured records to the same output
```

Comment:
isn't this the primary concern we have now with json records? co-locating structured and non-structured could lead to a large number of fields which ES could reject.

Reply:
Yes, but by adding `structured` and `scheme` fields (and setting `via-index-name` for old ES) we will direct JSON with differnet schema to different indices, so each index get a sane and consistent set of fields. 
We are relying on the user to identify the schema of different logs streams, so ES pipelines should still take measures to handle badly-behaved logs, but assuming the configuration is correct I'm hoping that will result in manageable indices.